### PR TITLE
[Rhol-265] Implement 'Matches Expression'

### PR DESCRIPTION
**Pull request by odeac**
## Overview
Implement matches expressions in Rholang

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/RHOL-265

TODO: 
* add normaliser tests that test the behaviour regarding locallyFree and connectiveUsed

### CODE REVIEW

File: models/src/main/scala/coop/rchain/models/rholang/implicits.scala [hunk 1]
```diff
@@ -316,6 +320,8 @@ object implicits {
         case EAndBody(EAnd(p1, p2))     => p1.get.connectiveUsed || p2.get.connectiveUsed
         case EOrBody(EOr(p1, p2))       => p1.get.connectiveUsed || p2.get.connectiveUsed
         case EMethodBody(e)             => e.connectiveUsed
+        case EMatchesBody(EMatches(process, pattern)) =>
+          process.get.connectiveUsed || pattern.get.connectiveUsed
```
Comment (goral09) [inline models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: You shouldn't have to call `.get` anymore. There're probably couple residues. There must be an implicit that goes from `Par => Option[Par]` that makes compiler accept those constructs.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -487,6 +488,19 @@ object Reduce {
             b1 <- evalToBool(p1.get)
             b2 <- evalToBool(p2.get)
           } yield GBool(b1 || b2)
+
+        case EMatchesBody(EMatches(target, pattern)) =>
+          for {
+            evaledTarget <- evalExpr(target.get)
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Here as well, I think those `.get`s are not necessary.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala [hunk 3]
```diff
@@ -894,4 +894,50 @@ class ReduceSpec extends FlatSpec with Matchers with PersistentStoreTester {
...
+      val reducer      = RholangOnlyDispatcher.create[Task, Task.Par](space).reducer
+
+      val inspectTask =  reducer.evalExpr(EMatches(GInt(1), EVar(Wildcard(Var.WildcardMsg()))))
+
+      Await.result(inspectTask.runAsync, 3.seconds)
+    }
+
+    result.exprs should be (Seq(Expr(GBool(true))))
+    errorLog.readAndClearErrorVector should be(Vector.empty[InterpreterError])
+  }
```
Comment (goral09) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: You should also test for `1 matches x` where `x` is some bound variable.

File: rholang/src/main/bnfc/rholang_mercury.cf [hunk 4]
```diff
@@ -43,7 +43,7 @@ PLt.          Proc7  ::= Proc7 "<" Proc8 ;
 PLte.         Proc7  ::= Proc7 "<=" Proc8 ;
 PGt.          Proc7  ::= Proc7 ">" Proc8 ;
 PGte.         Proc7  ::= Proc7 ">=" Proc8 ;
-PMatches.     Proc6  ::= Proc7 "matches" Proc7 ;
+PMatches.     Proc6  ::= Proc7 "matches" ProcVar ;
```
Comment (guardbotmk3) [inline rholang/src/main/bnfc/rholang_mercury.cf]: This is wrong. Why did you change this?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 5]
```diff
@@ -424,8 +424,9 @@ object ProcNormalizeMatcher {
       case p: PEq  => binaryExp(p.proc_1, p.proc_2, input, EEq.apply)
       case p: PNeq => binaryExp(p.proc_1, p.proc_2, input, ENeq.apply)
 
-      case p: PAnd => binaryExp(p.proc_1, p.proc_2, input, EAnd.apply)
-      case p: POr  => binaryExp(p.proc_1, p.proc_2, input, EOr.apply)
+      case p: PAnd     => binaryExp(p.proc_1, p.proc_2, input, EAnd.apply)
+      case p: POr      => binaryExp(p.proc_1, p.proc_2, input, EOr.apply)
+      case p: PMatches => binaryExp(p.proc_1, p.proc_2, input, EMatches.apply)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: This doesn't even match the grammar change that you made.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala (hunk 3 above)
Comment (guardbotmk3) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Do you mean `1 matches =x` where `x` is a variable bound to `1`

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala [hunk 6]
```diff
@@ -894,4 +894,50 @@ class ReduceSpec extends FlatSpec with Matchers with PersistentStoreTester {
...
+      val reducer      = RholangOnlyDispatcher.create[Task, Task.Par](space).reducer
+
+      val inspectTask =  reducer.evalExpr(EMatches(GInt(1), GInt(0)))
+
+      Await.result(inspectTask.runAsync, 3.seconds)
+    }
+
+    result.exprs should be (Seq(Expr(GBool(false))))
+    errorLog.readAndClearErrorVector should be(Vector.empty[InterpreterError])
+  }
```
Comment (guardbotmk3) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: These 2 examples are kind of trivial. It would be nice to see a more complicated example.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala (hunk 3 above)
Comment (goral09) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Yes, my mistake.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala (hunk 6 above)
Comment (odeac) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: The matching functionality is already tested with the tests for the `match-case` construction. 

What I'm testing here is that the matching algorithm is called when a `matches` expression is reduced.

Since I haven't done any changes to the matching functionality, adding more ReduceSpec tests for the `matches` operator doesn't improve the test coverage.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala (hunk 3 above)
Comment (odeac) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Please see 84d3b59b

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala (hunk 6 above)
Comment (odeac) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: The examples in 84d3b59 should cover the usage of BoundevVars and Connectives

File: models/src/main/scala/coop/rchain/models/rholang/implicits.scala (hunk 1 above)
Comment (odeac) [reply models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 7]
```diff
@@ -309,6 +309,8 @@ object Substitute {
             s2(par1.get, par2.get)(EAnd(_, _))
           case EOrBody(EOr(par1, par2)) =>
             s2(par1.get, par2.get)(EOr(_, _))
+          case EMatchesBody(EMatches(target, pattern)) =>
+            s2(target.get, pattern.get)(EMatches(_, _))
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: you don't need those `.get`s.

File: models/src/main/scala/coop/rchain/models/rholang/implicits.scala [hunk 8]
```diff
@@ -316,7 +320,9 @@ object implicits {
         case EAndBody(EAnd(p1, p2))     => p1.get.connectiveUsed || p2.get.connectiveUsed
         case EOrBody(EOr(p1, p2))       => p1.get.connectiveUsed || p2.get.connectiveUsed
         case EMethodBody(e)             => e.connectiveUsed
-        case ExprInstance.Empty         => false
+        case EMatchesBody(EMatches(process, pattern)) =>
+          process.connectiveUsed || pattern.connectiveUsed
```
Comment (guardbotmk3) [inline models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: No, this is still wrong.

File: models/src/main/scala/coop/rchain/models/rholang/implicits.scala [hunk 9]
```diff
@@ -347,7 +353,9 @@ object implicits {
         case EAndBody(EAnd(p1, p2))     => p1.get.locallyFree | p2.get.locallyFree
         case EOrBody(EOr(p1, p2))       => p1.get.locallyFree | p2.get.locallyFree
         case EMethodBody(e)             => e.locallyFree
-        case ExprInstance.Empty         => BitSet()
+        case EMatchesBody(EMatches(target, pattern)) =>
+          target.locallyFree | pattern.locallyFree
```
Comment (guardbotmk3) [inline models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: Same here.

File: models/src/main/scala/coop/rchain/models/rholang/implicits.scala (hunk 8 above)
Comment (goral09) [reply models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: So the `connectiveUsed` and `locallyFree` should be only calculated from the `process` (target) and not the `pattern` as well?

Comment (odeac) [reply models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: what would be a rholang test that will fail because of this?

File: models/src/main/scala/coop/rchain/models/rholang/implicits.scala (hunk 9 above)
Comment (odeac) [reply models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: what would be a rholang test that will fail because of this?
